proposals/p1787.md

```diff
@@ -0,0 +1,97 @@
+# Reorder type and length in array type syntax
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1787)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+The status quo syntax for an expression representing the type of an array has
+some slight annoyances and inconsistencies. This is certainly not a major issue,
+but it does impact users broadly; anyone writing multidimensional arrays.
+
+Specifically, the problem is that the array extents are in reverse order from
+how the array is indexed. That is, given an array of length 5, each element of
+which is an array of 10 `i32`s, consider a variable declaration along with an
+indexing into the "last" element.
+
+```
+var array: [[i32; 10]; 5]
+var element: array[4][9]
+```
+
+The concern is that users would be tempted to write the expression
+`array[9][4]`. More generally, the concern is that users writing multi-indexing
+will be tempted to assume the first index corresponds to the lexically first
+extent in the type when the proper correspondence is to the outermost (and
+hence, with the current syntax, lexically last) extent.
+
+## Background
+
+No background knowledge is required.
+
+## Proposal
+
+This concern can be addressed by simply reversing the order of the type and the
+extent syntactically so that the length appears first. Then the example above
+would be rewritten as
+
+```
+var array: [5; [10; i32]]
+var element: array[4][9]
+```
+
+## Details
+
+Because array indexing is read left-to-right, and each index peels off one layer
+of a multidimensional array, reading indexes left-to-right corresponds to
+indexing from the outermost to innermost extents. By listing the extents before
+the type syntactically, this outer-to-inner ordering becomes compatible with the
+left-to-right ordering.
+
+## Rationale
+
+This proposal addresses Carbon's goal to promote
+[Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write).
+It does not have any affect (positive or negative) on any of the other goals.
+
+## Alternatives considered
```

Comment:
Should we consider the alternative of supporting true multidimensional arrays (as opposed to arrays of arrays)?
I.e. multiple sizes can be provided inside the same brackets, one per dimension:

```zig
// declaration syntax could look something like this:
var array: [i32; 5 x 10 x 2]
var array: [i32; 5, 10, 2]

// indexing could look like this:
var element = array[4, 9, 1]
var element = array[4][9][1]
```

This would make the syntax even more lightweight.

The multidimensional subscript operator syntax `[x, y, z]` is the same that we have in C++23: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p2128r0.pdf

Reply:
I don't think this resolves the issue: Arrays of arrays would still be possible. Maybe they'd be discouraged? But we'd now have `[i32; 5, 10]` be more like `[[i32; 10]; 5]` which also seems backwards. Insofar as people can be confused by the backwardness of nested arrays where the length follows the type, I'd like to fix that.

I certainly would also be in favor syntactic sugar, so `[5, 10, 2; i32]` would be equivalent to `[5; [10; [2; i32]]]`, but I'd still want the extents to precede the types. I had thought about this sugar as well but wanted to keep this proposal focussed on just the ordering issue.

Reply:
I agree. If you're using the wrong abstraction, the sooner you realize it, the better.

That and the listed differences could be used as a reason for adding multidimensional arrays to the language. But I don't know a lot of languages that provide built-in multidimensional arrays (at least C# does, but I don't know any systems language that does). In my experience it's also very rare to see arrays of arrays in code. Most languages provide a multidimensional array class in a library for the relatively rare cases when it's needed. 

So based on those two points, I'm not sure if there's a strong enough incentive for providing built-in multidimensional arrays, or optimizing the array syntax for nested arrays. I think it makes more sense to optimize the array syntax for the common use cases, for C++ familiarity, for avoidance of [precedence issues](https://github.com/carbon-language/carbon-lang/pull/1787#discussion_r933322056), etc. A multidimensional array class can be easily provided in a library.

Reply:
The questions about multidimensional arrays are valid, but I think they're a [straw man](https://en.wikipedia.org/wiki/Straw_man). Whether we want to people to write nested arrays or not, they will. If we have multidimensional arrays, people will write arrays of multidimensional arrays or multidimensional arrays of multidimensional arrays. All of the extent/indexing concerns mentioned in this proposal still apply in those situations.

I think @emlai's prioritization list is roughly the correct one and match Carbon's goals regarding readability. I say "roughly" because I'm skeptical that familiarity is worth preserving, or even that reordering the extent and the type affects readability due to familiarity in a meaningful way.

Reply:
For more native support of multidimensional arrays, I would prefer the current syntax for array declaration. It is close to Fortran array definition. There are some examples:
```fortran
  real(8), dimension(4) :: EigVal
  real(8), dimension(4,4) :: EigVec
```
or
```fortran
  real(8) :: EigVal(4), EigVec(4,4)
```
In both cases, firstly we know the type of variables (type is more important then dimensions)
They are close to current declaration of arrays in Carbon:
```
  var EigVal: [f64, 4];
  var EigVec: [[f64, 4], 4];
```
And here, we know type before definition of dimensions.

> But I don't know a lot of languages that provide built-in multidimensional arrays (at least C# does, but I don't know any systems language that does). In my experience it's also very rare to see arrays of arrays in code. 

Native multidimensional arrays are main point why Fortran is continuing to use nowadays. It is actively using in mathematical applications for modelling. There is many things which are mostly interesting in 2D or 3D cases rather then in 1D. Also, (I come from Quantum Chemistry) we have matrixes of electron interactions most of them are 2D (I saw some codes which uses arrays up to 6D arrays for high-level theories).

> A multidimensional array class can be easily provided in a library.

I would prefer native implementation. C++ is much more progressive language than Fortran but its libraries supporting multidimensional arrays requires to write a lot of code for slicing that makes tensor contraction is not readable.

Reply:
> I don't know a lot of languages that provide built-in multidimensional arrays (at least C# does, but I don't know any systems language that does).

I'd argue that C and C++ provide built-in multidimensional arrays. For example, this is valid C/C++:

```c
int a[2][3] = {1, 2, 3, 4, 5, 6};
```

`{1, 2, 3, 4, 5, 6}` makes no sense as an initializer for an array of arrays of ints, but it does make sense as an initializer for an array of 6 ints.

As another example that's especially significant for this PR: the array type syntax is very definitely not nested. Consider:
```c++
using Array = int[3][4][5];
Array a = ...;
```
`a` is definitely not an array of 5 `int[3][4]`s. If it was, you could access the last `int[3][4]` in that array with `a[4]`, meaning that the index order would be reversed. C's decision to make the array declaration syntax match the indexing syntax _forces_ it to reject array nesting. In fact, C's declaration syntax is very cleverly constructed to stop you from writing a nested array declaration that specifies the size of more than one dimension.

Reply:
> They are close to current declaration of arrays in Carbon

This is what I'm proposing we change.

I think I agree that the element type is more important than the extents, but I don't have any way to quantify that importance. I'm also not sure that important things should necessarily be first. The slightly weaker claim I agree with is that they element type should be readily apparent. Because I expect the number of characters spent on extents to be low on average, I think the element type will be apparent.

@geoffromer I understood the syntactic claims you made but I think I'm missing the larger point you're making, and how it relates to multidimensional arrays. Also, are you making positive or negative value judgments on the suitability of C's syntax in Carbon? Or just observing? Thanks in advance for the clarification.

Reply:
> @geoffromer I understood the syntactic claims you made but I think I'm missing the larger point you're making, and how it relates to multidimensional arrays. Also, are you making positive or negative value judgments on the suitability of C's syntax in Carbon? Or just observing? Thanks in advance for the clarification.

I think it's noteworthy that C avoids the index-order problem for nested arrays by steering programmers away from nested arrays, which in broad strokes is what I'm arguing Carbon should do. However, C had very different goals in this area, and its "steering" is far more heavy-handed than Carbon's can or should be, so I don't think that's a very decisive precedent.

Other than that, I was just observing. Among other things, it took me quite a while to understand how C's multidimensional array type syntax works, and either this proposal or a closely-related one is going to need to address that as an alternative-considered, so I thought it might save you (or whoever) some work.